ddtrace/contrib/langchain/patch.py

```diff
@@ -0,0 +1,713 @@
+        for param, val in getattr(instance, "_identifying_params", {}).items():
+            if isinstance(val, dict):
+                for k, v in val.items():
+                    span.set_tag_str("langchain.request.%s.parameters.%s.%s" % (llm_provider, param, k), str(v))
+            else:
+                span.set_tag_str("langchain.request.%s.parameters.%s" % (llm_provider, param), str(val))
+
+        chat_completions = func(*args, **kwargs)
+        if isinstance(instance, langchain.chat_models.ChatOpenAI):
+            _tag_openai_token_usage(span, chat_completions.llm_output)
+            integration.record_usage(span, chat_completions.llm_output)
+
+        for message_set_idx, message_set in enumerate(chat_completions.generations):
+            for idx, chat_completion in enumerate(message_set):
+                if integration.is_pc_sampled_span(span):
+                    span.set_tag_str(
+                        "langchain.response.completions.%d.%d.content" % (message_set_idx, idx),
+                        integration.trunc(chat_completion.text),
+                    )
+                span.set_tag_str(
+                    "langchain.response.completions.%d.%d.message_type" % (message_set_idx, idx),
+                    chat_completion.message.__class__.__name__,
+                )
+
+        if integration.is_pc_sampled_log(span):
+            integration.log(
+                span,
+                "info" if span.error == 0 else "error",
+                "sampled %s.%s" % (instance.__module__, instance.__class__.__name__),
+                attrs={
+                    "messages": [
+                        [
+                            {
+                                "content": message.content,
+                                "message_type": message.__class__.__name__,
+                            }
+                            for message in messages
+                        ]
+                        for messages in chat_messages
+                    ],
+                    "choices": [
+                        [
+                            {
+                                "content": message.text,
+                                "message_type": message.message.__class__.__name__,
+                            }
+                            for message in messages
+                        ]
+                        for messages in chat_completions.generations
+                    ],
+                },
+            )
+    except Exception:
+        span.set_exc_info(*sys.exc_info())
+        integration.metric(span, "incr", "request.error", 1)
+        raise
+    finally:
+        span.finish()
+        integration.metric(span, "dist", "request.duration", span.duration_ns)
+    return chat_completions
+
+
+@with_traced_module
+async def traced_chat_model_agenerate(langchain, pin, func, instance, args, kwargs):
+    llm_provider = instance._llm_type.split("-")[0]
```

Comment:
There is a decent amount of duplication here with `traced_chat_model_generate` could we define more helpers to consolidate some of the logic. Is this worth the effort or would it negatively impact readability.

Reply:
This might actually negatively impact readability at this point because there are 3 sync/async pairs of methods we need to do this for and this might actually increase even more complexity at this point.